crates/binjs_es6/src/sublanguages.rs

```diff
@@ -0,0 +1,426 @@
+}
+impl PropertiesGuard {
+    fn new(stack: Rc<RefCell<Vec<Properties>>>, top: Properties) -> Self {
+        {
+            stack.borrow_mut().push(top);
+        }
+        PropertiesGuard { stack }
+    }
+
+    /// Create a new `PropertiesGuard` for a complex tree.
+    fn complex(visitor: &InjectVisitor, _path: &WalkPath) -> Self {
+        let top = Properties::Complex;
+        Self::new(visitor.stack.clone(), top)
+    }
+
+    /// Create a new `PropertiesGuard` for a pure data tree.
+    fn pure_data(visitor: &InjectVisitor, size: usize, _path: &WalkPath) -> Self {
+        let top = Properties::PureDataSoFar { size };
+        Self::new(visitor.stack.clone(), top)
+    }
+}
+impl Drop for PropertiesGuard {
+    /// When we drop the PropertiesGuard:
+    ///
+    /// - if both child and parent are `PureDataSoFar`, increase the number of nodes
+    ///   in the parent subtree;
+    /// - otherwise, mark the parent as `Complex`.
+    fn drop(&mut self) {
+        let mut stack = self.stack.borrow_mut();
+        let me = stack.pop().unwrap();
+
+        if stack.len() == 0 {
+            // We are the root, nothing to do.
+            return;
+        }
+
+        let parent = stack.last_mut().unwrap();
+        let new_parent = match (&me, &*parent) {
+            // If either the parent or the child is complex, the parent is complex.
+            (&Properties::Complex, _) | (_, &Properties::Complex) => Properties::Complex,
+
+            // If both are pure so far, add both sizes.
+            (
+                &Properties::PureDataSoFar { size: my_size },
+                &Properties::PureDataSoFar { size: parent_size },
+            ) => Properties::PureDataSoFar {
+                size: my_size + parent_size,
+            },
+        };
+        *parent = new_parent
+    }
+}
+
+impl WalkGuard<InjectVisitor> for PropertiesGuard {
+    /// Constructor called by default methods of `InjectVisitor`
+    fn new(origin: &InjectVisitor, path: &WalkPath) -> Self {
+        PropertiesGuard::complex(origin, path)
+    }
+}
+
+type EnterResult = Result<VisitMe<PropertiesGuard>, ()>;
+type ExitResult<T> = Result<Option<T>, ()>;
+
+impl InjectVisitor {
+    pub fn new(pure_data_threshold: usize) -> Self {
+        Self {
+            pure_data_threshold,
+            stack: Rc::new(RefCell::new(Vec::new())),
+        }
+    }
+    pub fn rewrite_script(pure_data_threshold: usize, script: &mut Script) {
+        let mut visitor = Self::new(pure_data_threshold);
+        script
+            .walk(&mut WalkPath::new(), &mut visitor)
+            .expect("Could not walk script");
+    }
+
+    /// The properties of the current node.
+    fn properties(&self) -> Properties {
+        self.stack.borrow().last().unwrap().clone()
+    }
+
+    /// The properties of the parent node, or `None` if there is no parent.
+    fn parent_properties(&self) -> Option<Properties> {
+        let borrow = self.stack.borrow();
```

Comment:
can this be renamed to something more descriptive?
IMO, `borrow` doesn't explain any about what it is.
can it be `stack` or something?